src/components/permissions/rowLevelPolicyHelpers.ts

```diff
@@ -0,0 +1,327 @@
+import databaseConnect from '../databaseConnect'
+import { UserInfo, PermissionRow, RuleTypes } from './types'
+
+const SQLBuilder = require('json-sql-builder2')
+const jsonToSql = new SQLBuilder('PostgreSQL')
+
+// in { permissionPolicyId: 1, permissionNameId:2 }
+// out "pp1pn2"
+const getPermissionNameAbbreviation = ({ permissionPolicyId, permissionNameId }: PermissionRow) =>
+  `pp${permissionPolicyId}pn${permissionNameId}`
+
+// in { permissionPolicyId: 1, permissionNameId:2, templatePermissionId: 3 }
+// out "pp1pn2tp3"
+const getTemplatePermissionAbbreviation = (permissionRow: PermissionRow) =>
+  `${getPermissionNameAbbreviation(permissionRow)}tp${permissionRow.templatePermissionId}`
+
+// Returns new version of object with each key prefixed with prefix (and value converter to string)
+// in "prefixMe", { one: 1, two: 2}
+// out { prefixMe_one: "1", prefixMe_two: "2" }
+const remapObjectKeysWithPrefix = (prefix: string, object: Object) => {
+  const remappedObject: { [index: string]: string } = {}
+
+  Object.entries(object).forEach(([key, value]) => {
+    remappedObject[`${prefix}_${key}`] = String(value)
+  })
+
+  return remappedObject
+}
+
+/* Compiles JWT from userInfo and PerissionRows
+  in { userId: 1, ... }, [
+  {
+    templatePermissionRestrictions: {
+      restrictOne: 1,
+      restrictTwo: 2,
+    },
+    templateId: 1,
+    permissionPolicyId: 2,
+    permissionNameId: 3,
+    templatePermissionId: 4,
+  }]
+  out {
+    aud: 'postgraphile',
+    userId: 1,
+    pp2pn3: true,
+    pp2pn3tp4: true,
+    pp2pn3tp4_templateId: "1",
+    pp2pn3tp4_restrictOne: "1",
+    pp2pn3tp4_restrictTwo: "2"
+  }
+*/
+const compileJWT = (userInfo: UserInfo, templatePermissionRows: Array<PermissionRow>) => {
+  let JWT = { userId: userInfo.userId, aud: 'postgraphile' }
+
+  templatePermissionRows.forEach((permissionRow: PermissionRow) => {
+    const { templatePermissionRestrictions, templateId, templatePermissionId } = permissionRow
+
+    let permissionAbbreviation = getPermissionNameAbbreviation(permissionRow)
+
+    JWT = {
+      [permissionAbbreviation]: true,
+      ...JWT,
+    }
+    // It's possible to have permission name given to user without any templatePermissionIds
+    if (!templatePermissionId) return // 'continue' syntax for for each
+
+    permissionAbbreviation = getTemplatePermissionAbbreviation(permissionRow)
+    JWT = {
+      [permissionAbbreviation]: true,
+      ...remapObjectKeysWithPrefix(permissionAbbreviation, {
+        ...templatePermissionRestrictions,
+        templateId,
+      }),
+      ...JWT,
+    }
+  })
+  return JWT
```

Comment:
Hmm, personally it's more readable for me in the for each format, what do you think ?
https://github.com/openmsupply/application-manager-server/blob/78363a9416d564bd2f78800b62bccf07e0ac8cc1/src/components/permissions/rowLevelPolicyHelpers.ts#L52-L77

Reply:
I liked more with the reduce, Idk. It's just more straight forward to me. Althought this bit inside is a little tricky to understand. I'm getting confused with the reuse of `permissionAbbreviation`. Maybe like this would be clearer:

```
const permissionAbbreviation = getPermissionNameAbbreviation(permissionRow) 

if (! templatePermissionId) 
  return { 
         [permissionAbbreviation]: true, 
         ...JWT, 
       }
else {
const templatePermissionAbbreviation = getTemplatePermissionAbbreviation(permissionRow)

   return {
         [permissionAbbreviation]: true,
         [templatePermissionAbbreviation]: true, 
         ...remapObjectKeysWithPrefix(permissionAbbreviation, { 
           ...templatePermissionRestrictions, 
           templateId, 
         }), 
         ...JWT, 
       } 
}
```

Is this the same as the above? I haven't tested this code.

Reply:
> Is this the same as the above? I haven't tested this code.

Pretty much `remapObjectKeysWithPrefix(permissionAbbreviation, { ` would need to have templatePermissionAbbreviation rather then permissionAbbreviation.

> I'm getting confused with the reuse of permissionAbbreviation

Interesting enough i had it like you have it previously (still in forEach).

Anyways, I am not convinced that reduce is more readable. If you feel strongly about please ask Carl to be the tie breaker. (sorry I would usually just make the requested change, but reduce feels a bit weird to me here, especially since the initial `{ userId, username, aud: 'postgraphile' } ` seems kind of hidden)

I've made more comments and removed reuse of permissionAbbreviation though.

c6876efc6868b292b066fa68490587a4d3f11143

Reply:
The change you had previously in forEach could have been kept in the `reduce` version to keep it simpler too. 
I still think it would look better, but I really don't think it's worth a big argument around this. Maybe is just a matter of preference.

I'm going to have another look on `rowLevelPolicyHelpers.ts ` I couldn't see clearly what the code is doing 2 steps ahead. That's why I wanted to help with small changes to see if it would be clearer. I always think that it could look simpler even though it's doing hardcore stuff.

Reply:
Cools, I was hoping we could reach a compromise with comments made in this commit https://github.com/openmsupply/application-manager-server/commit/c6876efc6868b292b066fa68490587a4d3f11143